---
output:
  html_document: default
  pdf_document: default
---

```{r}
library(Seurat)
library(ggplot2)
library(Rtsne)
library(rgl)
library(dplyr)
library(reshape)
library(RColorBrewer)
library(kohonen)
library(ComplexHeatmap)
library(circlize)
library(GetoptLong)
```

Completed table
```{r}
# load("Neftel 500 6000.RData")
# #Reading output from Differential Expression
# neftel.markers <-  readRDS("NeftelDElist.RDS")
# #Reading the outputfrom seurat cell cycle phase assignment
# neftel.cellcycle <- readRDS("NeftelCellScoringUPDATED.RDS")
exp.matFINAL <- read.table(file = "leblancmatrix2002000(scaled).txt", header = TRUE, as.is = TRUE, row.names = 1)
#View(neftel.markers)


# # Cut down expression table to just DE genes
# exp.matDE <- exp.matFINAL[rownames(exp.matFINAL) %in% leblanc.markers$gene,]
# #Obtain filtered cells and phases
# cellcycle.assignment<-data.frame(sort(neftel.cellcycle$Phase))
# #Cut down expression table by relevant cells only 
# exp.matDEfinal <- exp.matDE[colnames(exp.matDE) %in% rownames(cellcycle.assignment)]
# #Order cells by phase
# colnames(exp.matDEfinal) <- intersect(rownames(cellcycle.assignment),colnames(exp.matDEfinal))

```


```{r}
###############################################
#Average expression table and relevant plots

###############################################
load("Leblanc 200 2000.RData")

# Find the average expression of the genes in each phase
leblancavg <- AverageExpression(leblanc,slot = "scale.data") 

# Filter the average expression table down to just the DE genes
leblancavg <- as.data.frame(leblancavg)
length(intersect(rownames(leblancavg),leblanc.markers$gene)) #184
leblancavg <- leblancavg[rownames(leblancavg) %in% leblanc.markers$gene,]



###PCA CLUSTERING

leblancpca <- prcomp(leblancavg)

pca.scores <- data.frame(leblancpca$x)

leblanc.val <- cbind(leblancavg,pca.scores)
head(leblanc.val)

 
### SOM CLUSTERING

leblanc.val$gene <- rownames(leblanc.val)
leblanc.val <- leblanc.val %>% select(gene, everything())

som.data <- as.matrix(leblanc.val[,c(2:4)])
set.seed(2023)

som.data <- som(som.data, grid = somgrid(6, 6, "hexagonal"))
summary(som.data)

plot(som.data, main = "leblanc data",shape="straight")

# TACC3 cluster list
somclusterlist <- data.frame(rownames(leblanc.val),som.data$unit.classif)
grep("^23$",somclusterlist$som.data.unit.classif,value =T)
TACC3list.leblanc <- somclusterlist[grep("^23$",somclusterlist$som.data.unit.classif),]


## use hierarchical clustering to cluster 
som.data$codes <- as.data.frame(som.data$codes)
som_cluster <- cutree(hclust(dist(som.data$codes)),k = 5)
# plot these results:
plot(som.data, type="mapping", bgcol = som_cluster, main = "Clusters")
add.cluster.boundaries(som.data, som_cluster) 


### 3D scatter plot
#plot3d(leblancavg)
#rglwidget()

```

```{r}
#Dotplot of TACC3 cluster
DotPlot(leblanc, features = TACC3list.leblanc$rownames.leblanc.val. ,dot.scale = 8) + RotatedAxis()

```


SOM clustering on all genes
```{r}
leblancavgall <- AverageExpression(leblanc,slot = "scale.data") 
leblancavgall <- as.data.frame(leblancavgall)

#PCA CLUSTERING

leblancallpca <- prcomp(leblancavgall)

allpca.scores <- data.frame(leblancallpca$x)

leblanc.valall <- cbind(leblancavgall,allpca.scores)
head(leblanc.valall)

# SOM CLUSTERING

leblanc.valall$gene <- rownames(leblanc.valall)
leblanc.valall <- leblanc.valall %>% select(gene, everything())

som.data <- as.matrix(leblanc.valall[,c(2:4)])
set.seed(2023)
som.data <- som(som.data, grid = somgrid(6, 6, "hexagonal"))
summary(som.data)

plot(som.data, main = "leblanc data",shape="straight")

somclusterlist <- data.frame(rownames(leblanc.valall),som.data$unit.classif)
grep("^31$",somclusterlist$som.data.unit.classif,value =T)
TACC3listall.leblanc <- somclusterlist[grep("^31$",somclusterlist$som.data.unit.classif),]

## use hierarchical clustering to cluster the codebook vectors
som.data$codes <- as.data.frame(som.data$codes)
som_cluster <- cutree(hclust(dist(som.data$codes)),k = 5)
# plot these results:
plot(som.data, type="mapping", bgcol = som_cluster, main = "Clusters")
add.cluster.boundaries(som.data, som_cluster) 

length(intersect(TACC3list.leblanc$rownames.leblanc.val.,TACC3listall.leblanc$rownames.leblanc.valall.)) # 2 (half DE genes)

```


Complex Heatmap - SOM derived TACC3 list for average expression (leblanc) 
```{r}
exp.matFINAL <- read.table(file = "leblancmatrix2002000(scaled).txt", header = TRUE, as.is = TRUE, row.names = 1)
exp.matFINAL <- leblancexpmatFINAL


Subtype <- readRDS("LeblancSubtypeVIS.RDS")

#Loading the final expression matrix for scaled neftel data
#exp.matFINAL <- read.table(file = "", header = TRUE, as.is = TRUE, row.names = 1)

# Cut down expression table to just TACC3 related genes from SOM Clustering
Tacc3leblanc.mtx <- exp.matFINAL[rownames(exp.matFINAL) %in% TACC3list.leblanc$rownames.leblanc.val.,]
Tacc3leblanc.mtx <- exp.matFINAL[rownames(exp.matFINAL) %in% TACC3listall.leblanc$rownames.leblanc.valall.,]

Tacc3leblanc.mtx <- exp.matFINAL[rownames(exp.matFINAL) %in% overlapDE,]### 118 genes boxplots ###
Tacc3leblanc.mtx <- exp.matFINAL[rownames(exp.matFINAL) %in% unique(leblanc.markers$gene),] # leblanc DE genes

#Annotating columns with subtype
ha=HeatmapAnnotation(Subtype=Subtype$svm.subtype.call, col=list(Subtype=c("Classical"="#d8b365","Proneural"="#af8dc3","Mesenchymal" = "#5ab4ac")))
ha2=HeatmapAnnotation(Phase=leblanc$Phase, col=list(Phase=c("S"="#e41a1c","G2M"="#377eb8","G1"="#4daf4a")))


#Annotating columns with Cell Cycle Phase
#ha=HeatmapAnnotation(Phase=leblanc$Phase, col=list(Phase=c("S"="lightgreen","G2M"="lightblue","G1"="orange")))

#Constructing a ComplexHetamp
ht_list = Heatmap(data.matrix(Tacc3leblanc.mtx), col = colorRamp2(c(-1.5, 0, 1.5), c("blue", "white", "red")),border = "white" ,name = "scaled_expr",row_title = "Clustering of TACC3 related genes from SOM",show_column_names = FALSE,show_row_names = TRUE ,top_annotation = c(ha,ha2),row_names_max_width = unit(3, "cm"),width = unit(10, "cm"),na_col = "black",column_km =3,gap=unit(10,"mm"),row_names_gp = gpar(fontsize = 10),heatmap_legend_param = list(title = "Expression",use_raster = TRUE))
#save.image("/rds/projects/g/gendood-preclinomics/Bioinfo_Module_6_Group_Project/som_ht_neftel.RData")

#View the heatmap
ht_list

```

```{r}
colnames(Tacc3leblanc.mtx) <- Subtype$svm.subtype.call

# Calculate the row-wise average for each subtype
Tacc3leblanc.mtx$Classical.average <- rowMeans(Tacc3leblanc.mtx[, grepl("Classical", colnames(Tacc3leblanc.mtx))])
Tacc3leblanc.mtx$Mesenchymal.average <- rowMeans(Tacc3leblanc.mtx[, grepl("Mesenchymal", colnames(Tacc3leblanc.mtx))])
Tacc3leblanc.mtx$Proneural.average <- rowMeans(Tacc3leblanc.mtx[, grepl("Proneural", colnames(Tacc3leblanc.mtx))])

Tacc3.mtx.subset <- Tacc3leblanc.mtx[, c("Classical.average", "Mesenchymal.average", "Proneural.average")]


# Create a boxplot of the expression values for each subtype
boxplot(Tacc3.mtx.subset, main="Gene Expression by Subtype", names=c("Classical", "Mesenchymal", "Proneural"), ylab="Expression Value")

# boxplot with jitter
expression_data_long <- Tacc3.mtx.subset %>%
  rownames_to_column("Gene") %>%
  gather(Class, Expression, -Gene)

ggplot(expression_data_long, aes(x = Class, y = Expression)) +
  geom_boxplot(outlier.shape = NA) +
  labs(x = "Class", y = "Expression") +
  ggtitle("Gene Expression by Subtype") +
  geom_jitter()
  


# all DE genes
exp.matFINAL <- read.table(file = "leblancmatrix2002000(scaled).txt", header = TRUE, as.is = TRUE, row.names = 1)


length(intersect(rownames(exp.matFINAL),leblanc.markers$gene)) #184
exp.matFINAL <- exp.matFINAL[rownames(exp.matFINAL) %in% leblanc.markers$gene,]
#exp.matFINAL <- exp.matFINAL[intersect(colnames(exp.matFINAL),rownames(Subtype))]

colnames(exp.matFINAL) <- Subtype$svm.subtype.call

# Calculate the row-wise average for each subtype
exp.matFINAL$Classical.average <- rowMeans(exp.matFINAL[, grepl("Classical", colnames(exp.matFINAL))])
exp.matFINAL$Mesenchymal.average <- rowMeans(exp.matFINAL[, grepl("Mesenchymal", colnames(exp.matFINAL))])
exp.matFINAL$Proneural.average <- rowMeans(exp.matFINAL[, grepl("Proneural", colnames(exp.matFINAL))])

exp.matFINAL.subset <- exp.matFINAL[, c("Classical.average", "Mesenchymal.average", "Proneural.average")]


# Create a boxplot of the expression values for each subtype
boxplot(exp.matFINAL.subset, main="Gene Expression by Subtype", names=c("Classical", "Mesenchymal", "Proneural"), ylab="Expression Value")

# boxplot with jitter
expression_data_long <- exp.matFINAL.subset %>%
  rownames_to_column("Gene") %>%
  gather(Class, Expression, -Gene)

ggplot(expression_data_long, aes(x = Class, y = Expression)) +
  geom_boxplot(outlier.shape = NA) +
  labs(x = "Class", y = "Expression") +
  ggtitle("Gene Expression by Subtype") +
  geom_jitter()
  
### Phases boxplot ###

colnames(Tacc3leblanc.mtx) <- leblanc$Phase

# Calculate the row-wise average for each subtype
Tacc3leblanc.mtx$G2M.average <- rowMeans(Tacc3leblanc.mtx[, grepl("G2M", colnames(Tacc3leblanc.mtx))])
Tacc3leblanc.mtx$S.average <- rowMeans(Tacc3leblanc.mtx[, grepl("S", colnames(Tacc3leblanc.mtx))])
Tacc3leblanc.mtx$G1.average <- rowMeans(Tacc3leblanc.mtx[, grepl("G1", colnames(Tacc3leblanc.mtx))])


Tacc3.mtx.subset <- Tacc3leblanc.mtx[, c("G2M.average", "S.average","G1.average")]


# Create a boxplot of the expression values for each phase
boxplot(Tacc3.mtx.subset, main="Gene Expression by Subtype", names=c("G2M", "S", "G1"), ylab="Expression Value")

#DE genes
length(intersect(rownames(exp.matFINAL),leblanc.markers$gene)) #184
exp.matFINAL2 <- exp.matFINAL[rownames(exp.matFINAL) %in% leblanc.markers$gene,]

colnames(exp.matFINAL2) <- leblanc$Phase


# Calculate the row-wise average for each subtype
exp.matFINAL2$G2M.average <- rowMeans(exp.matFINAL2[, grepl("G2M", colnames(exp.matFINAL2))])
exp.matFINAL2$S.average <- rowMeans(exp.matFINAL2[, grepl("S", colnames(exp.matFINAL2))])
exp.matFINAL2$G1.average <- rowMeans(exp.matFINAL2[, grepl("G1", colnames(exp.matFINAL2))])

exp.matFINAL.subset <- exp.matFINAL2[, c("G2M.average", "S.average","G1.average")]

# Create a boxplot of the expression values for each phase
boxplot(exp.matFINAL.subset, main="Gene Expression by Subtype", names=c("G2M", "S", "G1"), ylab="Expression Value")
```

```{r}
# Create binary variables indicating high expression for each subtype based on the highest numerical value per row
Tacc3.mtx.subset_binary <- Tacc3.mtx.subset %>%
  mutate(Classical.binary = ifelse(Classical.average == pmax(Classical.average, Mesenchymal.average, Proneural.average), 1, 0),
         Mesenchymal.binary = ifelse(Mesenchymal.average == pmax(Classical.average, Mesenchymal.average, Proneural.average), 1, 0),
         Proneural.binary = ifelse(Proneural.average == pmax(Classical.average, Mesenchymal.average, Proneural.average), 1, 0))


# Create a 3x2 contingency table
#FIX THIS SOMETHING IS WRONG DOESNT ADD PROPERLY

contingency_table <- matrix(c(sum(Tacc3.mtx.subset_binary$Classical.binary),
                              sum(Tacc3.mtx.subset_binary$Mesenchymal.binary),
                              sum(Tacc3.mtx.subset_binary$Proneural.binary),
                              sum(!Tacc3.mtx.subset_binary$Classical.binary),
                              sum(!Tacc3.mtx.subset_binary$Mesenchymal.binary),
                              sum(!Tacc3.mtx.subset_binary$Proneural.binary)), nrow = 3, ncol = 2, byrow = F)

# Add row and column names to the contingency table
rownames(contingency_table) <- c("Classical", "Mesenchymal", "Proneural")
colnames(contingency_table) <- c("High", "Low")



# Perform Fisher's exact test
fisher_test <- fisher.test(contingency_table, hybrid = T)
fisher_test

# Perform Fisher's exact test
chisq_test <- chisq.test(contingency_table)
chisq_test

####################################################################################

# Create binary variables indicating high expression for each subtype based on the highest numerical value per row
Tacc3.mtx.subset_binary <- exp.matFINAL.subset %>%
  mutate(Classical.binary = ifelse(Classical.average == pmax(Classical.average, Mesenchymal.average, Proneural.average), 1, 0),
         Mesenchymal.binary = ifelse(Mesenchymal.average == pmax(Classical.average, Mesenchymal.average, Proneural.average), 1, 0),
         Proneural.binary = ifelse(Proneural.average == pmax(Classical.average, Mesenchymal.average, Proneural.average), 1, 0))


# Create a 3x2 contingency table
#FIX THIS SOMETHING IS WRONG DOESNT ADD PROPERLY

contingency_table <- matrix(c(sum(Tacc3.mtx.subset_binary$Classical.binary),
                              sum(Tacc3.mtx.subset_binary$Mesenchymal.binary),
                              sum(Tacc3.mtx.subset_binary$Proneural.binary),
                              sum(!Tacc3.mtx.subset_binary$Classical.binary),
                              sum(!Tacc3.mtx.subset_binary$Mesenchymal.binary),
                              sum(!Tacc3.mtx.subset_binary$Proneural.binary)), nrow = 3, ncol = 2, byrow = F)

# Add row and column names to the contingency table
rownames(contingency_table) <- c("Classical", "Mesenchymal", "Proneural")
colnames(contingency_table) <- c("High", "Low")



# Perform Fisher's exact test
fisher_test <- fisher.test(contingency_table, hybrid = T)
fisher_test

# Perform Fisher's exact test
chisq_test <- chisq.test(contingency_table)
chisq_test

```

Correlation matrix/T TEST
```{r}
cor(Tacc3.mtx.subset)
cor(exp.matFINAL.subset)

## T TEST

# Perform a t-test comparing Proneural with Classical group
ttest_classical_proneural <- t.test(Tacc3.mtx.subset$Proneural.average, Tacc3.mtx.subset$Classical.average)

# Perform a t-test comparing Proneural with Mesenchymal group
ttest_mesenchymal_proneural <- t.test(Tacc3.mtx.subset$Proneural.average, Tacc3.mtx.subset$Mesenchymal.average)

# Perform a t-test comparing Classical with Mesenchymal group
ttest_classical_mesenchymal <- t.test(Tacc3.mtx.subset$Classical.average, Tacc3.mtx.subset$Mesenchymal.average)

# Check the p-values of the t-tests
ttest_classical_proneural$p.value
ttest_mesenchymal_proneural$p.value
ttest_classical_mesenchymal$p.value




# Perform a t-test comparing Proneural with Classical group
ttest_classical_proneural <- t.test(exp.matFINAL.subset$Proneural.average, exp.matFINAL.subset$Classical.average)

# Perform a t-test comparing Proneural with Mesenchymal group
ttest_mesenchymal_proneural <- t.test(exp.matFINAL.subset$Proneural.average, exp.matFINAL.subset$Mesenchymal.average)

# Check the p-values of the t-tests
ttest_classical_proneural$p.value
ttest_mesenchymal_proneural$p.value

############# phases t test ############

# Perform a t-test comparing G2M with S group
ttest_G2M_S <- t.test(Tacc3.mtx.subset$G2M.average, Tacc3.mtx.subset$S.average)

# Perform a t-test comparing G2M with G1 group
ttest_G2M_G1 <- t.test(Tacc3.mtx.subset$G2M.average, Tacc3.mtx.subset$G1.average)

# Perform a t-test comparing S with G1 group
ttest_G1_S <- t.test(Tacc3.mtx.subset$G1.average, Tacc3.mtx.subset$S.average)

# Check the p-values of the t-tests
ttest_G2M_S$p.value
ttest_G2M_G1$p.value
ttest_G1_S$p.value

```


```{r}
leblancavg$G1_S_diff <- abs(leblancavg$RNA.G1 - leblancavg$RNA.S)
leblancavg$S_G2M_diff <- abs(leblancavg$RNA.S - leblancavg$RNA.G2M)
leblancavg$G1_G2M_diff <- abs(leblancavg$RNA.G1 - leblancavg$RNA.G2M)
leblancavg$mag_dif <- leblancavg$G1_S_diff + leblancavg$S_G2M_diff + leblancavg$G1_G2M_diff


hist(leblancavg$mag_dif)

summary(leblancavg$mag_dif)


#geneindex <- rownames(leblancavg)[leblancavg$mag_dif >  3.4698 ]
geneindexleblanc <- rownames(leblancavg)[leblancavg$mag_dif > quantile(leblancavg$mag_dif,0.7)]

#View(as.data.frame(geneindexleblanc))

length(intersect(geneindexleblanc,TACC3list.leblanc$rownames.leblanc.val.)) #2/4 SOM Genes, 
length(intersect(geneindexleblanc,TACC3listall.leblanc$rownames.leblanc.valall.)) #44/56 SOM genes

```

